pilot/pkg/serviceregistry/kube/controller/controller_test.go

```diff
@@ -1557,103 +1533,111 @@ func addNodes(t *testing.T, controller *Controller, nodes ...*coreV1.Node) {
-	if ev := fx.Wait("eds"); ev == nil {
-		t.Fatalf("Timeout incremental eds")
-	}
+			// Endpoints are generated by Kubernetes from pod labels and service selectors.
+			// Here we manually create them for mocking purpose.
+			svc1Ips := []string{"128.0.0.1"}
+			portNames := []string{"tcp-port"}
+			// Create 1 endpoint that refers to a pod in the same namespace.
+			createEndpoints(controller, "svc1", "nsa", portNames, svc1Ips, t)
+			if ev := fx.Wait("eds"); ev == nil {
+				t.Fatalf("Timeout incremental eds")
+			}
 
-	// delete normal service
-	err := controller.client.CoreV1().Services("nsa").Delete("svc1", &metaV1.DeleteOptions{})
-	if err != nil {
-		t.Fatalf("Cannot delete service (error: %v)", err)
-	}
-	if ev := fx.Wait("service"); ev == nil {
-		t.Fatalf("Timeout deleting service")
-	}
+			// delete normal service
+			err := controller.client.CoreV1().Services("nsa").Delete("svc1", &metaV1.DeleteOptions{})
+			if err != nil {
+				t.Fatalf("Cannot delete service (error: %v)", err)
+			}
+			if ev := fx.Wait("service"); ev == nil {
+				t.Fatalf("Timeout deleting service")
+			}
 
-	// 2. full xds push request for headless service endpoint update
+			// 2. full xds push request for headless service endpoint update
 
-	// create a headless service
-	createServiceWithoutClusterIP(controller, "svc1", "nsa", nil,
-		[]int32{8080}, map[string]string{"app": "prod-app"}, t)
-	if ev := fx.Wait("service"); ev == nil {
-		t.Fatalf("Timeout creating service")
-	}
+			// create a headless service
+			createServiceWithoutClusterIP(controller, "svc1", "nsa", nil,
+				[]int32{8080}, map[string]string{"app": "prod-app"}, t)
+			if ev := fx.Wait("service"); ev == nil {
+				t.Fatalf("Timeout creating service")
+			}
 
-	// Create 1 endpoint that refers to a pod in the same namespace.
-	svc1Ips = append(svc1Ips, "128.0.0.2")
-	updateEndpoints(controller, "svc1", "nsa", portNames, svc1Ips, t)
-	if ev := fx.Wait("xds"); ev == nil {
-		t.Fatalf("Timeout xds push")
+			// Create 1 endpoint that refers to a pod in the same namespace.
+			svc1Ips = append(svc1Ips, "128.0.0.2")
+			updateEndpoints(controller, "svc1", "nsa", portNames, svc1Ips, t)
+			if ev := fx.Wait("xds"); ev == nil {
+				t.Fatalf("Timeout xds push")
+			}
+		})
 	}
 }
 
 // Validates that when Pilot sees Endpoint before the corresponding Pod, it loads Pod from K8S and proceed.
 func TestEndpointUpdateBeforePodUpdate(t *testing.T) {
-	// Setup kube caches
-	controller, fx := newFakeController()
-	defer controller.Stop()
-	pod1 := generatePod("172.0.1.1", "pod1", "nsA", "", "node1", map[string]string{"app": "prod-app"}, map[string]string{})
-	pod2 := generatePod("172.0.1.2", "pod2", "nsA", "", "node2", map[string]string{"app": "prod-app"}, map[string]string{})
+	for mode, name := range EndpointModeNames {
+		t.Run(name, func(t *testing.T) {
+			controller, fx := newFakeControllerWithOptions(fakeControllerOptions{mode: mode})
```

Comment:
I think this is not right, you need to copy the value of mode. otherwise, if the subtests are run in parrallel, when the function is run, the mode may be the last value of `EndpointModeNames `

Reply:
FYI https://gist.github.com/posener/92a55c4cd441fc5e5e85f27bca008721

Reply:
I think it doesn't apply here since mode isn't a pointer? I'll change it though since I'm not sure

Reply:
I put a hold on this to fix it. Once this is in I'll update to beta endpoint slice api

Reply:
It doesnot matter whether it is value type or pointer type, as mode is a variable

Reply:
I have fixed this, please take a look